### ISSUE TITLE
Inconsistent entity at DB: geo-located attribute but not location field

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#3167]
Sometimes CB reports error like this one:

```
"statusCode": {
                "code": "472",
                "reasonPhrase": "request parameter is invalid/not allowed",
                "details": "action: APPEND - entity: [E1, T] - offending attribute: position - attempt to define a geo location attribute [position] when another one has been previously defined []"
```

This is the reference request that were used to raise that error:

```
{
    "url": "http://localhost:1026/v1/updateContext",
    "method": "POST",
    "headers": {
        "fiware-service": "sc_smart_region_andalucia",
        "fiware-servicepath": "/and_sr_motril",
        "X-Auth-Token": null
    },
    "json": {
        "contextElements": [
            {
                "type": "T",
                "isPattern": "false",
                "id": "E1",
                "attributes": [
                    {
                        "name": "measure1",
                        "type": "string",
                        "value": "44",
                        "metadatas": [
                            {
                                "name": "TimeInstant",
                                "type": "ISO8601",
                                "value": "2018-03-20T16:58:31.421Z"
                            }
                        ]
                    },
                    {
                        "name": "measure2",
                        "type": "string",
                        "value": "21348",
                        "metadatas": [
                            {
                                "name": "TimeInstant",
                                "type": "ISO8601",
                                "value": "2018-03-20T16:58:31.421Z"
                            }
                        ]
                    },
                    {
                        "value": "32.7121, -2.39172",
                        "type": "geo:point",
                        "name": "position",
                        "metadatas": [
                            {
                                "name": "TimeInstant",
                                "type": "ISO8601",
                                "value": "2018-03-20T16:58:31.421Z"
                            }
                        ]
                    },
                    {
                        "name": "TimeInstant",
                        "type": "ISO8601",
                        "value": "2018-03-20T16:58:31.421Z"
                    }
                ]
            }
        ],
        "updateAction": "APPEND"
    }
}
```

The error message corresponds to a checking that CB does and it is correct in normal circustance. But what is abnormal (and buggy) in this case is the `...defined []` part, without specifying an actual attribute.

Looking into DB we have the following:

```
﻿{
    "_id" : {
        "id" : "E1",
        "type" : "T",
        "servicePath" : "/subserv1"
    },
    "attrNames" : [
        "TimeInstant",
        "position",
        "measure1",
        "measure2",
        "measure3"
    ],
    "attrs" : {
        "TimeInstant" : {
            "value" : 1521565051.0,
            "type" : "ISO8601",
            "mdNames" : [],
            "creDate" : 1521194680,
            "modDate" : 1521565051
        },
        "position" : {
            "type" : "geo:point",
            "creDate" : 1521194680,
            "modDate" : 1521194680,
            "value" : "36.7141, -4.47172",
            "mdNames" : []
        },
        "measure1" : {
            "value" : "22",
            "type" : "string",
            "md" : {
                "TimeInstant" : {
                    "type" : "ISO8601",
                    "value" : "2018-03-20T16:57:31.502Z"
                }
            },
            "mdNames" : [
                "TimeInstant"
            ],
            "creDate" : 1521194680,
            "modDate" : 1521565051
        },
        "measure2" : {
            "value" : "...",
            "type" : "string",
            "md" : {
                "TimeInstant" : {
                    "type" : "ISO8601",
                    "value" : "2018-03-20T16:57:31.502Z"
                }
            },
            "mdNames" : [
                "TimeInstant"
            ],
            "creDate" : 1521565051,
            "modDate" : 1521565051
        },
        "measure3" : {
            "value" : ...,
            "type" : "Number",
            "md" : {
               "TimeInstant" : {
                    "type" : "ISO8601",
                    "value" : "2018-03-20T15:00:00.289Z"
                }
            },
            "mdNames" : [
                "TimeInstant"
            ],
            "creDate" : 1521194680,
            "modDate" : 1521558000
        }
    },
    "creDate" : 1521193681,
    "modDate" : 1521565052,
    "lastCorrelator" : "d21fea80-2c5f-44e8-be07-fa163e734aab"
}
```

That document is inconsistent, as it has a location attribute:

```
        "position" : {
            "type" : "geo:point",
            "creDate" : 1521194680,
            "modDate" : 1521194680,
            "value" : "32.7121, -2.39172",
            "mdNames" : []
        },
```

but it doesn has the corresponding `location` field (see https://fiware-orion.readthedocs.io/en/master/admin/database_model/index.html#entities-collection):

```
   "location": {
       "attrName": "position",
       "coords": {
           "type": "Point",
           "coordinates": [ -2.39172, 32.7121 ]
       }
   },
```

My guess is that the root cause of the `...defined []` problem is in the update logic, that causes Orion to end with an inconsistent entity in the DB. However, it is unknown at the present moment how to put Orion in such state (i.e. the request of sequence of requests that causes an entity with the inconsistency in the DB). Once reproduced, the fix should be easy.

CC: @alvarosainzpardo

**Comment by fgalan** [comment]
Maybe related https://github.com/telefonicaid/fiware-orion/issues/1142 ?

**Comment by arigliano** [comment]
Hello, I would like to work on this issue. Can you assign it to me?

**Comment by fgalan** [comment]
Done. Thanks!

**Comment by fgalan** [comment]
This issue could have been closed by PR https://github.com/telefonicaid/fiware-orion/pull/3283, but a .test should confirm it before closing.

**Comment by fgalan** [comment]
In fact, I think no new .test is needed. Using the .test already included in PR #3283 running with a 1.15.0-next version (i.e. previous to the fix implemented in PR #3283) I have been able to reproduce the problem.

In particular, replace_location_attribute_with_same_location_attribute.test ends with the following status in the database:

```
> db.entities.find().pretty()
{
	"_id" : {
		"id" : "E1",
		"type" : "T1",
		"servicePath" : "/"
	},
	"attrNames" : [
		"A1"
	],
	"attrs" : {
		"A1" : {
			"type" : "geo:point",
			"creDate" : 1536858285,
			"modDate" : 1536858285,
			"value" : "2, -4",
			"mdNames" : [ ]
		},
		"location" : {
			"attrName" : "A1",
			"coords" : {
				"type" : "Point",
				"coordinates" : [
					-4,
					2
				]
			}
		}
	},
	"creDate" : 1536858285,
	"modDate" : 1536858285,
	"location" : {
		"attrName" : "A1",
		"coords" : {
			"type" : "Point",
			"coordinates" : [
				2,
				-5
			]
		}
	},
	"lastCorrelator" : "1d8e4762-b777-11e8-9c5f-000c29173617"
}
```

which is inconsistent, as we have a attrs.location key with an inconsistent value.

In addition, with replace_no_location_attribute_with_location_attribute.test the situation at the end in database is:

```
> db.entities.find().pretty()
{
	"_id" : {
		"id" : "E1",
		"type" : "T1",
		"servicePath" : "/"
	},
	"attrNames" : [
		"A1"
	],
	"attrs" : {
		"A1" : {
			"type" : "geo:point",
			"creDate" : 1536858362,
			"modDate" : 1536858362,
			"value" : "2, -4",
			"mdNames" : [ ]
		},
		"location" : {
			"attrName" : "A1",
			"coords" : {
				"type" : "Point",
				"coordinates" : [
					-4,
					2
				]
			}
		}
	},
	"creDate" : 1536858362,
	"modDate" : 1536858362,
	"lastCorrelator" : "4b14371e-b777-11e8-978d-000c29173617"
}
```

which is exaclty the same described in this issue body.

Thanks @arigliano for the fix. It wasn't easy to reproduce ;)
